File: tests/fixtures/SsrfToctouVulnerable.cs
using System.IO;
using System.Net.Http;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;

internal sealed class SsrfToctouVulnerable
{
    private readonly HttpClient client = new HttpClient();

    internal Task<HttpResponseMessage> FetchPreview(HttpRequest request)
    {
        string target = request.Query["url"].ToString();
        return client.GetAsync(target);
    }

    internal string ReadExisting(string path)
    {
        if (File.Exists(path))
        {
            return File.ReadAllText(path);
        }
        return string.Empty;
    }

    internal async Task<string> ReadExistingAsync(string path)
    {
        if (File.Exists(path))
        {
            return await File.ReadAllTextAsync(path);
        }
        return string.Empty;
    }
}

File: tests/fixtures/ssrf_toctou_vulnerable.go
package fixtures

import (
	"net/http"
	"os"
)

func fetchPreview(request *http.Request) (*http.Response, error) {
	target := request.URL.Query().Get("url")
	return http.Get(target)
}

func readExisting(path string) ([]byte, error) {
	if _, err := os.Stat(path); err == nil {
		return os.ReadFile(path)
	}
	return nil, os.ErrNotExist
}

func readAfterGuard(path string) ([]byte, error) {
	if _, err := os.Stat(path); err != nil {
		return nil, err
	}
	return os.ReadFile(path)
}
